src/lxc/usernsexec.c

```diff
@@ -0,0 +1,451 @@
+			continue;
+
+		p1 = strchr(line, ':');
+		if (!p1)
+			continue;
+
+		p2 = strchr(p1 + 1, ':');
+		if (!p2)
+			continue;
+
+		line[strlen(line) - 1] = '\0';
+		*p2 = '\0';
+
+		ret = lxc_safe_ulong(p1 + 1, &ul1);
+		if (ret < 0)
+			break;
+
+		ret = lxc_safe_ulong(p2 + 1, &ul2);
+		if (ret < 0)
+			break;
+
+		ret = add_map_entry(active_map, ul1, 0, ul2, which);
+		break;
+	}
+
+	return ret;
+}
+
+static int find_default_map(struct list_head *active_map)
+{
+	__do_free char *buf = NULL;
+	ssize_t bufsize;
+	struct passwd pwent;
+	int ret = -1;
+	struct passwd *pwentp = NULL;
+
+	bufsize = sysconf(_SC_GETPW_R_SIZE_MAX);
+	if (bufsize < 0)
+		bufsize = 1024;
+
+	buf = malloc(bufsize);
+	if (!buf)
+		return -1;
+
+	ret = getpwuid_r(getuid(), &pwent, buf, bufsize, &pwentp);
+	if (!pwentp) {
+		if (ret == 0)
+			CMD_SYSERROR("Failed to find matched password record");
+
+		CMD_SYSERROR("Failed to get password record for uid %d", getuid());
+		return -1;
+	}
+
+	ret = read_default_map(active_map, subuidfile, ID_TYPE_UID, pwent.pw_name);
+	if (ret < 0)
+		return -1;
+
+	ret = read_default_map(active_map, subgidfile, ID_TYPE_GID, pwent.pw_name);
+	if (ret < 0)
+		return -1;
+
+	return 0;
+}
+
+static bool is_in_ns_range(unsigned long id, struct id_map *map)
+{
+	if (id < map->nsid)
+		return false;
+
+	if (id >= map->nsid + map->range)
+		return false;
+
+	return true;
+}
+
+static bool do_map_self(struct list_head *active_map)
+{
+	struct id_map *map;
+	long nsuid = 0, nsgid = 0;
+	int ret;
+
+	list_for_each_entry(map, active_map, head) {
+		if (map->idtype == ID_TYPE_UID) {
+			if (is_in_ns_range(nsuid, map))
+				nsuid += map->range;
+		} else {
+			if (is_in_ns_range(nsgid, map))
+				nsgid += map->range;
+		}
+	}
+
+	ret = add_map_entry(active_map, getgid(), nsgid, 1, ID_TYPE_GID);
+	if (ret < 0)
+		return false;
+	ret = add_map_entry(active_map, getuid(), nsuid, 1, ID_TYPE_UID);
+	if (ret < 0)
+		return false;
+	return true;
+}
+
+__hidden extern int lxc_log_fd;
+
+int lxc_usernsexec(int argc, char *argv[], int *status)
+{
+	int c, pid, ret;
+	char buf[1];
+	int pipe_fds1[2], /* child tells parent it has unshared */
+	    pipe_fds2[2]; /* parent tells child it is mapped and may proceed */
+	unsigned long flags = CLONE_NEWUSER | CLONE_NEWNS;
+	char ttyname0[256] = {0}, ttyname1[256] = {0}, ttyname2[256] = {0};
+	char *default_args[] = {"/bin/sh", NULL};
+	bool map_self = false;
+	LIST_HEAD(active_map);
+
+
+	lxc_log_fd = STDERR_FILENO;
+
+	if (isatty(STDIN_FILENO)) {
+		ret = readlink("/proc/self/fd/0", ttyname0, sizeof(ttyname0));
+		if (ret < 0) {
+			CMD_SYSERROR("Failed to open stdin");
+			return 1;
+		}
+
+		ret = readlink("/proc/self/fd/1", ttyname1, sizeof(ttyname1));
+		if (ret < 0) {
+			CMD_SYSINFO("Failed to open stdout. Continuing");
+			ttyname1[0] = '\0';
+		}
+
+		ret = readlink("/proc/self/fd/2", ttyname2, sizeof(ttyname2));
+		if (ret < 0) {
+			CMD_SYSINFO("Failed to open stderr. Continuing");
+			ttyname2[0] = '\0';
+		}
+	}
```

Comment:
For example, I have zero idea why we actually need to do this and why this is even relevant for this binary. If we can't figure out why we should kick this.